core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -116,6 +117,10 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
       log.log(Level.INFO, "Received headers on closed stream {0} {1}",
           new Object[]{id(), headers});
     }
+    if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
+      String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
+      setDecompressor(messageEncoding);
```

Comment:
@ejona86 : question about this line:  is it safe to throw an exception from here?   If the stream has access to the listener, should the stream be closed, the listener contacted, and this method complete normally?